lms/services/exceptions.py

```diff
@@ -91,6 +91,49 @@ def __str__(self):
         return repr(self)
 
 
+class ExternalAsyncRequestError(ExternalRequestError):
+    def __init__(
+        self,
+        message=None,
+        response=None,
+        exception: Optional[Exception] = None,
+    ):
+        if exception and message:
+            message = f"{message}: {exception.__class__.__name__}"
```

Comment:
I like the idea of reporting the original exception that caused the `ExternalAsyncRequestError`. But:

1. I think we can already access the original exception as `self.__cause__` so we don't need the `exception` argument
2. I think it's good to keep `__cause__` and `message` as two separate attributes, rather than combining them into one string
3. I think I'd like to add `__cause__` support to `ExternalRequestError` so that `ExternalRequestError` and `ExternalAsyncRequestError` remain the same and so that all request errors have `__cause__` reporting not just the async ones

It think we could add `__cause__` to `ExternalRequestError.__repr__()` (which appears in Papertrail) and to [the `external_request_error()` exception view](https://github.com/hypothesis/lms/blob/594cb2793597578df82222f88b62ab777441c4a6/lms/views/api/exceptions.py#L111-L153) (which controls what appears in Sentry and in the error dialog).

For example currently if a request to the Canvas API times out `ExternalRequestError` doesn't make it clear that the problem was a timeout. If you hack the code like this:

```diff
diff --git a/lms/services/canvas_api/_basic.py b/lms/services/canvas_api/_basic.py
index ec641864..3eebf12b 100644
--- a/lms/services/canvas_api/_basic.py
+++ b/lms/services/canvas_api/_basic.py
@@ -99,6 +99,7 @@ class BasicClient:
         )
 
     def _send_prepared(self, request, schema, timeout, request_depth=1):
+        request.url = "https://www.doesntexist.com/"
         response = None
 
         try:
```

You'll get an error dialog like this:

![Screenshot from 2022-01-20 17-58-48](https://user-images.githubusercontent.com/22498/150396513-b321359f-b65d-4142-a36f-2f957350f92a.png)

If we add `__cause__` handling like this:

```diff
diff --git a/lms/views/api/exceptions.py b/lms/views/api/exceptions.py
index f6d78591..184d0004 100644
--- a/lms/views/api/exceptions.py
+++ b/lms/views/api/exceptions.py
@@ -140,6 +140,7 @@ class APIExceptionViews:
         return self.error_response(
             message=message,
             details={
+                "cause": str(self.context.__cause__),
                 "request": {
                     "method": self.context.method,
                     "url": strip_queryparams(self.context.url),
```

then the cause will be shown in the error dialog:

![Screenshot from 2022-01-20 18-11-57](https://user-images.githubusercontent.com/22498/150397063-801a851f-7b78-4430-bb8b-710363ed879b.png)

We should also add `sentry_sdk.set_context()` calls for `__cause__` like we have for `request`, `response` and `validation_errors` so that it appears in Sentry, and we should add `__cause__` to `ExternalRequestError.__repr__()` so that it appears in Papertrail.

Reply:
`__cause__` handled in a different PR